OTAnalytics/plugin_prototypes/track_visualization/track_viz.py

```diff
@@ -122,6 +132,91 @@ def get(self) -> dict[str, str]:
         return self._palette
 
 
+class EventToFlowResolver:
+    def __init__(self, flow_repository: FlowRepository) -> None:
+        self._flow_repository = flow_repository
+
+    def resolve(self, events: Iterable[Event]) -> Iterable[FlowId]:
+        flow_ids: set[FlowId] = set()
+        for event in events:
+            flow_ids.update(self._resolve_flow_id_for(event.section_id))
+
+        return flow_ids
+
+    def _resolve_flow_id_for(self, section_id: Optional[SectionId]) -> set[FlowId]:
+        if section_id:
+            return set(
+                [
+                    flow.id
+                    for flow in self._flow_repository.flows_using_section(section_id)
+                ]
+            )
+        return set()
+
+
+class FlowLayerPlotter(DynamicLayersPlotter[FlowId], FlowListObserver):
+    def __init__(
+        self,
+        plotter_factory: EntityPlotterFactory,
+        flow_state: FlowState,
+        flow_repository: FlowRepository,
+        track_repository: TrackRepository,
+        event_repository: EventRepository,
+        flow_id_resolver: EventToFlowResolver,
+    ) -> None:
+        super().__init__(plotter_factory, flow_state.selected_flows, self.get_flow_ids)
+
+        flow_repository.register_flows_observer(self)
+        flow_repository.register_flow_changed_observer(self.notify_flow)
+        track_repository.observers.register(self.notify_invalidate)
+        event_repository.register_observer(self.notify_events)
+
+        self._repository = flow_repository
+        self._flow_id_resolver = flow_id_resolver
+
+    def notify_flow(self, flow: FlowId) -> None:
+        self.notify_layers_changed([flow])
+
+    def notify_flows(self, flows: list[FlowId]) -> None:
+        self.notify_layers_changed(flows)
+
+    def notify_events(self, events: EventRepositoryEvent) -> None:
+        flows_to_add = self._flow_id_resolver.resolve(events.added)
+        self._handle_add_update(flows_to_add)
+        flows_to_remove = self._flow_id_resolver.resolve(events.removed)
+        self._handle_remove(flows_to_remove)
```

Comment:
Just serving as a reminder of what we talked about today. The invalidation of the `DynamicLayersPlotter` on filter visualization changes needs to be implemented, too.

Reply:
> Events that were removed indicate the clearance of the event repository, right? But at no point in time do `EventRepositoryEvent`s imply the removal of `Flow`s. In that case, shouldn't `self._handle_app_update` be called instead of `self._handle_remove` to invalidate the cache of the flow plot layers?

That is correct. Events that were removed indicate that events for at least one section are removed from the repository. This is triggered for removed sections or changes to sections. Sections can only be removed when no flow is starting or ending at them. Thus, this indicates that flows starting or ending at the changed section need to be updated. `_handle_add_update` will reuse the existing plotter and invalidate its cache, while `_handle_remove` will remove the plotter. As soon as it is plotted again, the plotter is recreated. `_handle_add_update` seems to be the better option in this case.